File: benchmarks/bench_fpga_eth.py
import socket
import time
import struct
import argparse
import os

def run_benchmark():
    parser = argparse.ArgumentParser(description="XIIRegexBuilder Ethernet Benchmark")
    parser.add_argument("--fpga-ip", default="192.168.2.10")
    parser.add_argument("--fpga-port", type=int, default=7777)
    parser.add_argument("--host-port", type=int, default=7778)
    parser.add_argument("--input", default="inputs/large_test_strings.txt")
    parser.add_argument("--regex", default="inputs/regexes.txt")
    parser.add_argument("--window", type=int, default=16)
    parser.add_argument("--verify", action="store_true")
    args = parser.parse_args()

    # Load strings
    with open(args.input, 'r') as f:
        strings = [line.strip() for line in f if line.strip() and not line.startswith('#')]
    
    num_regex = 0
    with open(args.regex, 'r') as f:
        num_regex = sum(1 for line in f if line.strip() and not line.startswith('#'))

    print(f"Loaded {len(strings)} strings and {num_regex} regexes.")

    # UDP Setup
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind(('', args.host_port))
    sock.settimeout(1.0)

    # Metrics
    start_wall = time.perf_counter()
    total_matches = 0
    
    # Packet protocol constants
    # Header: SEQ(4) | NUM_STR(2)
    # Body: [LEN(1) | DATA(N)] x NUM
    
    seq_num = 0
    results = {}
    
    print(f"Starting Ethernet Benchmark against {args.fpga_ip}...")

    # For Phase 4/5 demonstration, we pack strings 1-per-packet to test RTT
    # In a full run, we would batch them.
    for s in strings:
        payload = struct.pack(">IH", seq_num, 1) # Header
        payload += struct.pack("B", len(s)) + s.encode('ascii') # Body
        
        sock.sendto(payload, (args.fpga_ip, args.fpga_port))
        
        try:
            data, addr = sock.recvfrom(2048)
            # Response: SEQ(4) | NUM(2) | STAMP(4) | [MATCH(bits) | HITS]
            # Since we sent 1 string, we expect 1 result.
            res_seq, res_num, res_stamp = struct.unpack(">IHI", data[:10])
            results[res_seq] = data[10:]
            seq_num += 1
        except socket.timeout:
            print(f"Timeout on seq {seq_num}")
            break

    end_wall = time.perf_counter()
    duration = end_wall - start_wall
    
    print("\n=== Ethernet Benchmark Summary ===")
    print(f"Total Time: {duration:.4f} seconds")
    print(f"Throughput: {len(results)/duration:.2f} strings/sec")
    print(f"Success: {len(results)}/{len(strings)}")
    
    sock.close()

if __name__ == "__main__":
    run_benchmark()

File: benchmarks/bench_fpga_eth.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <cstring>
#include <chrono>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <unistd.h>
#include "protocol.h"

int main(int argc, char* argv[]) {
    std::string inputFile = (argc > 1) ? argv[1] : "inputs/large_test_strings.txt";
    
    std::ifstream in(inputFile);
    std::vector<std::string> strings;
    std::string line;
    while (std::getline(in, line)) {
        if (line.empty() || line[0] == '#') continue;
        strings.push_back(line);
    }
    in.close();

    int sock = socket(AF_INET, SOCK_DGRAM, 0);
    struct sockaddr_in fpga_addr;
    memset(&fpga_addr, 0, sizeof(fpga_addr));
    fpga_addr.sin_family = AF_INET;
    fpga_addr.sin_port = htons(xiir::protocol::RX_PORT);
    inet_pton(AF_INET, xiir::protocol::FPGA_IP, &fpga_addr.sin_addr);

    struct timeval tv;
    tv.tv_sec = 1;
    tv.tv_usec = 0;
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

    char buffer[65535];
    uint32_t seq = 0;
    uint32_t total_strings_processed = 0;

    std::cout << "Starting C++ Ethernet Benchmark (" << strings.size() << " strings)..." << std::endl;
    auto start = std::chrono::high_resolution_clock::now();

    size_t batch_size = 16;
    for (size_t i = 0; i < strings.size(); i += batch_size) {
        size_t chunk_end = std::min(i + batch_size, strings.size());
        uint16_t chunk_len = (uint16_t)(chunk_end - i);

        // Build Request Packet
        // Layout: SEQ(4) | NUM_STR(2) | [LEN(1) | DATA(N)] x NUM_STR
        uint8_t pkt[1500];
        uint32_t seq_be = htonl(seq);
        uint16_t num_be = htons(chunk_len);
        
        memcpy(pkt + 0, &seq_be, 4);
        memcpy(pkt + 4, &num_be, 2);
        
        size_t offset = xiir::protocol::REQ_HEADER_SIZE;
        for (size_t j = i; j < chunk_end; ++j) {
            pkt[offset++] = (uint8_t)strings[j].length();
            memcpy(pkt + offset, strings[j].c_str(), strings[j].length());
            offset += strings[j].length();
        }

        sendto(sock, pkt, offset, 0, (struct sockaddr*)&fpga_addr, sizeof(fpga_addr));

        // Receive Response Packet
        // Layout: SEQ(4) | NUM_RES(2) | STAMP(4) | [MATCH(bits) | HITS] x NUM_RES
        int n = recv(sock, buffer, sizeof(buffer), 0);
        if (n < 0) {
            std::cerr << "Timeout on seq " << seq << std::endl;
            break;
        }

        if (n < (int)xiir::protocol::RES_HEADER_SIZE) {
            std::cerr << "Malformed response (too short): " << n << " bytes" << std::endl;
            break;
        }
        
        // Parse response header
        uint32_t res_seq;
        uint16_t res_num;
        uint32_t res_stamp;
        
        memcpy(&res_seq,   buffer + 0, 4);
        memcpy(&res_num,   buffer + 4, 2);
        memcpy(&res_stamp, buffer + 6, 4);
        
        res_seq   = ntohl(res_seq);
        res_num   = ntohs(res_num);
        res_stamp = ntohl(res_stamp);

        if (res_seq != seq) {
            std::cerr << "Sequence mismatch! Expected " << seq << ", got " << res_seq << std::endl;
        }

        total_strings_processed += res_num;
        seq++;
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff = end - start;

    std::cout << "\n=== C++ Ethernet Benchmark Summary ===" << std::endl;
    std::cout << "Total Time: " << diff.count() << " seconds" << std::endl;
    std::cout << "Throughput: " << total_strings_processed / diff.count() << " strings/sec" << std::endl;
    std::cout << "Success: " << total_strings_processed << "/" << strings.size() << std::endl;

    close(sock);
    return 0;
}
